### PR TITLE
REFACTOR-#4970: Rewrite base implementations of a partition' `width/length`

### LINKED ISSUE #4970: REFACTOR: rewrite base implementations of a partition' `width/length`
These functions can be reused for `pandas_on_python` execution.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4970 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/core/dataframe/pandas/partitioning/partition.py [hunk 1]
```diff
@@ -298,10 +298,7 @@ def length(self):
             The length of the object.
         """
         if self._length_cache is None:
-            cls = type(self)
-            func = cls._length_extraction_fn()
-            preprocessed_func = cls.preprocess_func(func)
-            self._length_cache = self.apply(preprocessed_func)
+            self._length_cache = self.apply(self._length_extraction_fn()).get()
```
Comment (pyrito) [inline modin/core/dataframe/pandas/partitioning/partition.py:301]: Ha, this is interesting. Doesn't `self._length_extraction_fn()` effectively do `type(self)._length_extraction_fn()`?

File: modin/core/execution/python/implementations/pandas_on_python/partitioning/partition.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ def get(self):
         Since this object is a simple wrapper, just return the copy of data.
         """
         self.drain_call_queue()
-        return self._data.copy()
+        return self._data.copy() if hasattr(self._data, "copy") else self._data
```
Comment (pyrito) [inline modin/core/execution/python/implementations/pandas_on_python/partitioning/partition.py:66]: In what case would `data` not have the `copy` attribute?

File: modin/core/dataframe/pandas/partitioning/partition.py (hunk 1 above)
Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition.py:301]: Almost, applying to an object, not a class, the search space should be slightly larger, but this does not bother us in this case. Therefore, I do not think that we need to do an additional action to calculate the class.

**Review by pyrito — APPROVED** [review]

**Comment by anmyachev** [comment]
@mvashishtha could you take a look?

**Review by mvashishtha — APPROVED** [review]

**Comment by anmyachev** [comment]
@mvashishtha we can merge now, I think

File: modin/core/dataframe/pandas/partitioning/partition.py (hunk 1 above)
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition.py:301]: If some execution relies on this implementation of `length`/`width` and at the same time has `preprocess_func`, which is useful for perf, then we may have perf penalties here.

Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition.py:301]: If I understand correctly, it helps if we apply this function several times in a row, but this is not the case. Also, if it's really needed, it's better to do it in the `apply` function itself.

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition.py:301]: > If I understand correctly, it helps if we apply this function several times in a row, but this is not the case. Also, if it's really needed, it's better to do it in the `apply` function itself.

for the current engines, yes

> Also, if it's really needed, it's better to do it in the apply function itself.

not sure if it is better in the apply

anyway, let's leave as is for now

**Pull request merged**